src/algebra/ordered_monoid.lean

```diff
@@ -598,7 +598,7 @@ instance with_top.canonically_ordered_add_monoid : canonically_ordered_add_monoi
 end canonically_ordered_add_monoid
 
 /-- A canonically linear-ordered additive monoid is a canonically ordered additive monoid
-    whose ordering is a decidable linear order. -/
+    whose ordering is a linear order. -/
 @[protect_proj]
 class canonically_linear_ordered_add_monoid (α : Type*)
```

Comment:
Do we want a multiplicative version of this? Are there any other missing multiplicative / additive versions?

Reply:
This is an interesting question! Mathematically, a canonically ordered add_monoid is something which looks like "the stuff which is >= 0" -- of course `nat` is used everywhere, and `nnreal` is used everywhere in measure thory. A canonically ordered monoid is something which looks like "the stuff which is >= 1" and these things seem to be much less useful in practice. The only example I know in mathlib is `pnat`, and this is not canonically ordered wrt to the usual ordering, because it's not true in `pnat` that x <= y iff there's some c such that y=cx (at least for the usual ordering). 

Aah! Actually this _might_ be true for ideals of a Dedekind domain! They're not linearly ordered (I really can't think of any examples of a canonically linearly ordered monoid other than (rationals >= 1) and (reals >= 1) and other stuff like this, which don't seem to be useful classes -- indeed in valuation theory the interval [0,1] is useful but this is somehow anticanonically linearly ordered: a <= b iff there exists c such that a=c*b). However this canonically linearly ordered class is only used once in mathlib so one wonders how useful this stuff is in general.

I'll add canonically ordered monoids (which will be a small amount of work but I can now see the use of it), and I may as well do the linear ones (for completeness, and because it will be easy to do). Thanks.

Reply:
OK so I sat down to do this yesterday, and discovered that it is *not* true that ideals of a Dedekind domain satisfy this! Consider the integers. Every ideal of the integers is principal, and of the form (n), i.e. generated by a natural number n, and (n) <= (m) iff n is in (m) iff n is a multiple of m iff there exists c such that n = c * m iff there exists (c) such that (n) = (c) * (m). So again the inequality goes the wrong way!

Conclusion: I know of no class in mathlib which is a `canonically_linear_ordered_monoid`. However I know of several classes ([0,1], ideals in a Dedekind domain) are totally ordered monoids and for which the _dual_ order makes them a `canonically_linear_ordered_monoid`.

I am now very unclear about whether adding this new class is a good idea. I'll ask on Zulip.

Reply:
Oh OK so now I have one example! So it's back on.

Reply:
(Zulip links: [1](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/There.20are.20no.20.60canonically_linear_ordered_monoid.60s) [2](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/Expressing.20.E2.84.95.20as.20a.20category.20and.20Fibonacci.20as.20a.20functor))